### PR TITLE
[Refactor] Support progressive test with fewer memory cost

### PR DESCRIPTION
**Pull request by sennnnn**
# Support progressive test

## Calculate Metrics

The current test tool require a middle list type variable to keep all of the model predict masks. When there are large number of test samples, many bugs will occur. We add `dataset.pre_eval` function which can avoid keeping all of the model prediction masks and only keep the pre-evaluation results.  So, we just need to use `pre_eval` to calculate regular metrics, like, `IoU`, `Dice`. 

`Cityscapes` metric may need to keep middle model predict masks. We make a judgement for `Cityscapes` to ensure keeping middle predict masks for metric calculation.

## Format Results

We also provide a progressive mode for `format_results`. But the `dataset.format_results` must support `indices` argument.

ps:

usage:

```bash
python tools/test.py [config_path] [ckpt_path] --eval mIoU # Default progressive test.
python tools/test.py [config_path] [ckpt_path] --eval cityscapes # Test on formatting results.
python tools/test.py [config_path] [ckpt_path] --format-only # Default progressive test.
```

### mmseg/apis/test.py

1. Deprecate `efficient_test` argument; (`efficient_test` is still valid but the return value of test api will be prediction save paths)
2. Add `pre_eval` argument; (for progressive metric calculation)
3. Add `format_only` argument; (for progressive results format)
4. Add `format_args` argument; (for progressive results format)

### tools/test.py

1. Add `eval_on_format_results` middle variable for `cityscapes` metric;
2. Deprecate `efficient_test` argument;

### mmseg/core/evaluation/eval_hooks.py

1. Deprecated `efficient_test` argument;
2. Add `pre_eval` argument to control evaluation action of test api when training;

### CODE REVIEW

File: mmseg/apis/test.py [hunk 1]
```diff
@@ -163,3 +165,149 @@ def multi_gpu_test(model,
     else:
         results = collect_results_cpu(results, len(dataset), tmpdir)
     return results
+
+
+def progressive_single_gpu_test(model,
```
Comment (xvjiarui) [inline mmseg/apis/test.py:170]:
```suggestion
def single_gpu_test_eval(model,
```

File: mmseg/apis/test.py [hunk 2]
```diff
@@ -163,3 +165,149 @@ def multi_gpu_test(model,
...
+    pixel_count_matrix = [
+        total_area_intersect, total_area_union, total_area_pred_label,
+        total_area_label
+    ]
+    # collect results from all ranks
+    if gpu_collect:
+        results = collect_count_results_gpu(pixel_count_matrix, 4 * world_size)
+    else:
+        results = collect_count_results_cpu(pixel_count_matrix, 4 * world_size,
+                                            tmpdir)
```
Comment (xvjiarui) [inline mmseg/apis/test.py:292]: Consider use `dist.all_reduce` to sync the results.

File: mmseg/apis/test.py [hunk 3]
```diff
@@ -163,3 +165,149 @@ def multi_gpu_test(model,
...
+
+    return total_area_intersect, total_area_union, total_area_pred_label, \
+        total_area_label
+
+
+# TODO: Support distributed test api
+def progressive_multi_gpu_test(model,
+                               data_loader,
+                               tmpdir=None,
+                               gpu_collect=False):
```
Comment (xvjiarui) [inline mmseg/apis/test.py:244]: May also consider GPU test to see if it is faster. 
Please also note whether it would take too much memory.

File: mmseg/datasets/custom.py [hunk 4]
```diff
@@ -303,6 +311,82 @@ def get_palette_for_custom_classes(self, class_names, palette=None):
...
+        ret_metrics_class.pop('Class', None)
+        for key, value in ret_metrics_class.items():
+            eval_results.update({
+                key + '.' + str(name): value[idx] / 100.0
+                for idx, name in enumerate(class_names)
+            })
+
+        if mmcv.is_list_of(results, str):
+            for file_name in results:
+                os.remove(file_name)
```
Comment (xvjiarui) [inline mmseg/datasets/custom.py:387]: Not needed.

File: tools/test.py [hunk 5]
```diff
@@ -134,20 +137,36 @@ def main():
...
+                                      args.opacity)
     else:
         model = MMDistributedDataParallel(
             model.cuda(),
             device_ids=[torch.cuda.current_device()],
             broadcast_buffers=False)
-        outputs = multi_gpu_test(model, data_loader, args.tmpdir,
-                                 args.gpu_collect, efficient_test)
+        if only_pixel_count:
+            outputs = progressive_multi_gpu_test(model, data_loader,
```
Comment (xvjiarui) [inline tools/test.py:165]: Only use progressive when `args.eval=True`

File: mmseg/apis/test.py [hunk 6]
```diff
@@ -120,46 +113,167 @@ def multi_gpu_test(model,
...
     # collect results from all ranks
     if gpu_collect:
-        results = collect_results_gpu(results, len(dataset))
+        processor = collect_processors_gpu(processor)
+    else:
+        processor = collect_processors_cpu(processor, tmpdir)
+    return processor
+
+
+def collect_processors_gpu(processor):
```
Comment (xvjiarui) [inline mmseg/apis/test.py:178]: The same as `mmcv.engine.collect_results_gpu()`?

File: mmseg/apis/test.py [hunk 7]
```diff
@@ -120,46 +113,167 @@ def multi_gpu_test(model,
...
+                recv[:shape[0]].cpu().numpy().tobytes())
+            # When data is severely insufficient, an empty part_result
+            # on a certain gpu could makes the overall outputs empty.
+            if part_processor:
+                sub_processors.append(part_processor)
+        main_processor.merge(sub_processors)
+        return main_processor
+
+
+def collect_processors_cpu(processor, tmpdir=None):
```
Comment (xvjiarui) [inline mmseg/apis/test.py:224]: The same as `mmcv.engine.collect_results_cpu()`?

File: mmseg/datasets/custom.py [hunk 8]
```diff
@@ -227,17 +224,24 @@ def prepare_test_img(self, idx):
...
+        """Get ground truth segmentation map by index for evaluation."""
+        if not isinstance(indexes, Iterable):
+            indexes = [indexes]
+        gt_seg_maps = []
+        for index in indexes:
+            seg_map = osp.join(self.ann_dir,
+                               self.img_infos[index]['ann']['seg_map'])
+            gt_seg_maps.append(
+                mmcv.imread(seg_map, flag='unchanged', backend='pillow'))
         return gt_seg_maps
```
Comment (xvjiarui) [inline mmseg/datasets/custom.py:246]: https://docs.python.org/3/library/itertools.html#itertools.islice

Comment (xvjiarui) [reply mmseg/datasets/custom.py:246]: Something like `islice(rank, len(dataset), world_size)` should handle samples on different gpus.

File: tools/deploy_test.py [hunk 9]
```diff
@@ -227,24 +227,26 @@ def main():
...
         if args.out:
             print(f'\nwriting results to {args.out}')
-            mmcv.dump(outputs, args.out)
+            mmcv.dump(processor.retrieval(), args.out)
         kwargs = {} if args.eval_options is None else args.eval_options
         if args.format_only:
-            dataset.format_results(outputs, **kwargs)
+            assert middle_save, 'When `middle_save` is True, the '
+            '`--format-only` is valid.'
+            dataset.format_results(processor.retrieval(), **kwargs)
```
Comment (xvjiarui) [inline tools/deploy_test.py:247]: We may change this to progressive just like eval. 
(do it later).

File: mmseg/datasets/custom.py [hunk 10]
```diff
@@ -227,18 +226,25 @@ def prepare_test_img(self, idx):
...
-        for img_info in self.img_infos:
-            seg_map = osp.join(self.ann_dir, img_info['ann']['seg_map'])
-            if efficient_test:
-                gt_seg_map = seg_map
-            else:
-                gt_seg_map = mmcv.imread(
-                    seg_map, flag='unchanged', backend='pillow')
-            gt_seg_maps.append(gt_seg_map)
-        return gt_seg_maps
+    def pre_eval(self, preds, indices):
```
Comment (xvjiarui) [inline mmseg/datasets/custom.py:229]: Docstring

File: mmseg/apis/test.py [hunk 11]
```diff
@@ -124,42 +84,37 @@ def multi_gpu_test(model,
...
     Returns:
-        list: The prediction results.
+        list: evaluation preparetion results.
     """
-
     model.eval()
-    results = []
+    pre_eval_results = []
     dataset = data_loader.dataset
+
```
Comment (xvjiarui) [inline mmseg/apis/test.py:94]: Add comments about why we are using `batch_sampler`.

File: mmseg/core/evaluation/eval_hooks.py [hunk 12]
```diff
@@ -13,31 +13,56 @@ class EvalHook(_EvalHook):
...
+            runner.log_buffer.output[name] = val
+        runner.log_buffer.ready = True
+
+        if self.save_best is not None:
+            if self.key_indicator == 'auto':
+                # infer from eval_results
+                self._init_rule(self.rule, list(eval_res.keys())[0])
+            return eval_res[self.key_indicator]
+
+        return None
```
Comment (xvjiarui) [inline mmseg/core/evaluation/eval_hooks.py:52]: Move these part to other PR.

File: mmseg/core/evaluation/eval_hooks.py [hunk 13]
```diff
@@ -13,31 +13,56 @@ class EvalHook(_EvalHook):
...
             Default: False.
-        efficient_test (bool): Whether save the results as local numpy files to
-            save CPU memory during evaluation. Default: False.
     Returns:
         list: The prediction results.
     """
 
     greater_keys = ['mIoU', 'mAcc', 'aAcc']
 
-    def __init__(self, *args, by_epoch=False, efficient_test=False, **kwargs):
```
Comment (xvjiarui) [inline mmseg/core/evaluation/eval_hooks.py:24]: Keep the API the same, add deprecate warning.

File: mmseg/apis/test.py [hunk 14]
```diff
@@ -1,68 +1,52 @@
...
-
     model.eval()
-    results = []
+    pre_eval_results = []
     dataset = data_loader.dataset
     prog_bar = mmcv.ProgressBar(len(dataset))
-    if efficient_test:
-        mmcv.mkdir_or_exist('.efficient_test')
-    for i, data in enumerate(data_loader):
+
```
Comment (xvjiarui) [inline mmseg/apis/test.py:38]: Add explainnation.

File: tools/test.py [hunk 15]
```diff
@@ -133,32 +134,29 @@ def main():
...
             broadcast_buffers=False)
-        outputs = multi_gpu_test(model, data_loader, args.tmpdir,
-                                 args.gpu_collect, efficient_test)
+        pre_eval_results = multi_gpu_test(model, data_loader, args.format_only,
+                                          eval_args, args.tmpdir,
+                                          args.gpu_collect)
 
     rank, _ = get_dist_info()
     if rank == 0:
-        if args.out:
```
Comment (xvjiarui) [inline tools/test.py:154]: Deprecate warning.

File: tests/test_eval_hook.py [hunk 16]
```diff
@@ -112,11 +112,7 @@ def test_epoch_eval_hook():
                                                       logger=runner.logger)
 
 
-def multi_gpu_test(model,
-                   data_loader,
-                   tmpdir=None,
-                   gpu_collect=False,
-                   efficient_test=False):
+def multi_gpu_test(model, data_loader, tmpdir=None, gpu_collect=False):
```
[the remark this answers is not in the document]
Comment (sennnnn) [reply tests/test_eval_hook.py:115]: Add deprecated warning and modify it to compact old version.

File: mmseg/core/evaluation/eval_hooks.py (hunk 12 above)
Comment (sennnnn) [reply mmseg/core/evaluation/eval_hooks.py]: Already remove

File: mmseg/core/evaluation/eval_hooks.py (hunk 13 above)
Comment (sennnnn) [reply mmseg/core/evaluation/eval_hooks.py:24]: Add deprecated warning and modify it to compact old version.

**Comment by xvjiarui** [comment]
- [ ] Documentation
- [ ] Unittest.

**Comment by xvjiarui** [comment]
After documentation and unittest are completed, please let mmdet3d test whether there is BC breaking issues.

File: docs_zh-CN/inference.md [hunk 17]
```diff
@@ -20,11 +20,11 @@ python tools/test.py ${配置文件} ${检查点文件} [--out ${结果文件}]
 
 可选参数:
 
-- `RESULT_FILE`: pickle 格式的输出结果的文件名，如果不专门指定，结果将不会被专门保存成文件。
+- `RESULT_FILE`: pickle 格式的输出结果的文件名，如果不专门指定，结果将不会被专门保存成文件。（MMseg v0.17 之后，args.out 将只会保存评估时的中间结果或者是分割图的保存路径。）
 - `EVAL_METRICS`: 在结果里将被评估的指标。这主要取决于数据集，  `mIoU`  对于所有数据集都可获得，像 Cityscapes 数据集可以通过 `cityscapes` 命令来专门评估，就像标准的 `mIoU`一样。
 - `--show`: 如果被指定，分割结果将会在一张图像里画出来并且在另一个窗口展示。它仅仅是用来调试与可视化，并且仅针对单卡 GPU 测试。请确认 GUI 在您的环境里可用，否则您也许会遇到报错 `cannot connect to X server`
 - `--show-dir`: 如果被指定，分割结果将会在一张图像里画出来并且保存在指定文件夹里。它仅仅是用来调试与可视化，并且仅针对单卡GPU测试。使用该参数时，您的环境不需要 GUI。
-- `--eval-options`: 评估时的可选参数，当设置 `efficient_test=True` 时，它将会保存中间结果至本地文件里以节约 CPU 内存。请确认您本地硬盘有足够的存储空间（大于20GB）。
+- `--eval-options`: 评估时的可选参数，当设置 `efficient_test=True` 时，它将会保存中间结果至本地文件里以节约 CPU 内存。请确认您本地硬盘有足够的存储空间（大于20GB）。（MMseg v0.17 之后，`efficient_test` 不再被支持，我们重构了 test api，通过使用一种渐近式的方式来提升评估和保存结果的效率。）
```
Comment (xvjiarui) [inline docs_zh-CN/inference.md:27]:
```suggestion
- `--eval-options`: 评估时的可选参数，当设置 `efficient_test=True` 时，它将会保存中间结果至本地文件里以节约 CPU 内存。请确认您本地硬盘有足够的存储空间（大于20GB）。（MMseg v0.17 之后，`efficient_test` 不再生效，我们重构了 test api，通过使用一种渐近式的方式来提升评估和保存结果的效率。）
```

**Comment by DCNSW** [comment]
MMdet3d can be supported by this pr successfully. No BC breaking issues.

File: mmseg/apis/train.py [hunk 18]
```diff
@@ -105,6 +105,8 @@ def train_segmentor(model,
             dist=distributed,
             shuffle=False)
         eval_cfg = cfg.get('evaluation', {})
+        eval_cfg['pre_eval'] = cfg.runner[
+            'pre_eval'] if 'pre_eval' in cfg.runner else False
```
Comment (xvjiarui) [inline mmseg/apis/train.py:109]: Why we need to get `pre_eval` form `cfg.runner`?

File: mmseg/core/evaluation/eval_hooks.py [hunk 19]
```diff
@@ -21,9 +24,20 @@ class EvalHook(_EvalHook):
...
+    def __init__(self,
+                 *args,
+                 pre_eval=False,
```
Comment (xvjiarui) [inline mmseg/core/evaluation/eval_hooks.py:29]: Move `pre_eval` behind `efficient_test` for BC

File: mmseg/core/evaluation/eval_hooks.py [hunk 20]
```diff
@@ -57,9 +71,20 @@ class DistEvalHook(_DistEvalHook):
...
```
Comment (xvjiarui) [inline mmseg/core/evaluation/eval_hooks.py:76]: Move pre_eval behind efficient_test for BC

File: tests/test_apis/test_single_gpu.py [hunk 21]
```diff
@@ -0,0 +1,72 @@
...
+    data_loader = DataLoader(
+        test_dataset,
+        batch_size=1,
+        sampler=None,
+        num_workers=0,
+        shuffle=False,
+    )
+    model = ExampleModel()
+
+    # Test efficient test compatibility (deprecated)
```
Comment (xvjiarui) [inline tests/test_apis/test_single_gpu.py:45]:
```suggestion
    # Test efficient test compatibility (will be deprecated)
```

File: mmseg/apis/train.py (hunk 18 above)
Comment (sennnnn) [reply mmseg/apis/train.py:109]: MMDet3d use train api with eval hook when training. However, the dataset of mmdet3d doesn't implement `pre_eval` function. So, we set pre_eval False in default when using training api and use schedule to set pre_eval True.

File: mmseg/apis/train.py [hunk 22]
```diff
@@ -106,6 +106,8 @@ def train_segmentor(model,
             dist=distributed,
             shuffle=False)
         eval_cfg = cfg.get('evaluation', {})
+        eval_cfg['pre_eval'] = cfg.evaluation[
```
Comment (xvjiarui) [inline mmseg/apis/train.py:109]: Not needed, `pre_eval=False` by default.

**Review by xvjiarui — APPROVED** [review]
